File: src/core/gfx/video_processor.cc
#include "core/gfx/video_processor.h"

#include "core/common/errors.h"

namespace wincap {

void VideoProcessor::Init(ID3D11Device5*        device,
                          ID3D11DeviceContext4* context,
                          UINT                  width,
                          UINT                  height,
                          DXGI_FORMAT           output_format) {
    width_         = width;
    height_        = height;
    output_format_ = output_format;

    WINCAP_THROW_IF_FAILED("video_processor", device->QueryInterface(IID_PPV_ARGS(video_device_.GetAddressOf())));
    WINCAP_THROW_IF_FAILED("video_processor", context->QueryInterface(IID_PPV_ARGS(video_context_.GetAddressOf())));

    D3D11_VIDEO_PROCESSOR_CONTENT_DESC desc{};
    desc.InputFrameFormat            = D3D11_VIDEO_FRAME_FORMAT_PROGRESSIVE;
    desc.InputFrameRate.Numerator    = 60;
    desc.InputFrameRate.Denominator  = 1;
    desc.InputWidth                  = width;
    desc.InputHeight                 = height;
    desc.OutputFrameRate.Numerator   = 60;
    desc.OutputFrameRate.Denominator = 1;
    desc.OutputWidth                 = width;
    desc.OutputHeight                = height;
    desc.Usage                       = D3D11_VIDEO_USAGE_PLAYBACK_NORMAL;

    WINCAP_THROW_IF_FAILED("video_processor",
        video_device_->CreateVideoProcessorEnumerator(&desc, enumerator_.GetAddressOf()));
    WINCAP_THROW_IF_FAILED("video_processor",
        video_device_->CreateVideoProcessor(enumerator_.Get(), 0, processor_.GetAddressOf()));

    // Use Rec.709 (SDR). For HDR (P010) we'd configure BT.2020 + PQ here.
    D3D11_VIDEO_PROCESSOR_COLOR_SPACE in_cs{};
    in_cs.RGB_Range          = 0;        // 0..255 (full range)
    in_cs.YCbCr_Matrix       = 1;        // BT.709
    in_cs.YCbCr_xvYCC        = 0;
    in_cs.Nominal_Range      = D3D11_VIDEO_PROCESSOR_NOMINAL_RANGE_0_255;
    video_context_->VideoProcessorSetStreamColorSpace(processor_.Get(), 0, &in_cs);

    D3D11_VIDEO_PROCESSOR_COLOR_SPACE out_cs{};
    out_cs.RGB_Range     = 0;
    out_cs.YCbCr_Matrix  = 1;
    out_cs.Nominal_Range = D3D11_VIDEO_PROCESSOR_NOMINAL_RANGE_16_235; // studio range YUV
    video_context_->VideoProcessorSetOutputColorSpace(processor_.Get(), &out_cs);

    RECT full{0, 0, static_cast<LONG>(width), static_cast<LONG>(height)};
    video_context_->VideoProcessorSetStreamSourceRect(processor_.Get(), 0, TRUE, &full);
    video_context_->VideoProcessorSetStreamDestRect(processor_.Get(),  0, TRUE, &full);
    video_context_->VideoProcessorSetOutputTargetRect(processor_.Get(), TRUE, &full);
}

void VideoProcessor::Convert(ID3D11Texture2D* src, ID3D11Texture2D* dest) {
    if (!processor_ || !src || !dest) return;

    Microsoft::WRL::ComPtr<ID3D11VideoProcessorInputView> in_view;
    D3D11_VIDEO_PROCESSOR_INPUT_VIEW_DESC in_desc{};
    in_desc.FourCC         = 0;
    in_desc.ViewDimension  = D3D11_VPIV_DIMENSION_TEXTURE2D;
    in_desc.Texture2D.MipSlice  = 0;
    in_desc.Texture2D.ArraySlice = 0;
    WINCAP_THROW_IF_FAILED("video_processor",
        video_device_->CreateVideoProcessorInputView(src, enumerator_.Get(), &in_desc,
            in_view.GetAddressOf()));

    Microsoft::WRL::ComPtr<ID3D11VideoProcessorOutputView> out_view;
    D3D11_VIDEO_PROCESSOR_OUTPUT_VIEW_DESC out_desc{};
    out_desc.ViewDimension = D3D11_VPOV_DIMENSION_TEXTURE2D;
    out_desc.Texture2D.MipSlice = 0;
    WINCAP_THROW_IF_FAILED("video_processor",
        video_device_->CreateVideoProcessorOutputView(dest, enumerator_.Get(), &out_desc,
            out_view.GetAddressOf()));

    D3D11_VIDEO_PROCESSOR_STREAM stream{};
    stream.Enable          = TRUE;
    stream.OutputIndex     = 0;
    stream.InputFrameOrField = 0;
    stream.PastFrames      = 0;
    stream.FutureFrames    = 0;
    stream.pInputSurface   = in_view.Get();

    WINCAP_THROW_IF_FAILED("video_processor",
        video_context_->VideoProcessorBlt(processor_.Get(), out_view.Get(), 0, 1, &stream));
}

void VideoProcessor::Reset() noexcept {
    processor_.Reset();
    enumerator_.Reset();
    video_context_.Reset();
    video_device_.Reset();
    width_ = height_ = 0;
}

} // namespace wincap

File: crates/wincap-core/src/video_processor.rs
use windows::core::Interface;
use windows::Win32::Foundation::RECT;
use windows::Win32::Graphics::Direct3D11::*;
use windows::Win32::Graphics::Dxgi::Common::{DXGI_FORMAT, DXGI_RATIONAL};

use crate::error::{hr_call, WincapResult};

#[derive(Clone, Copy, PartialEq, Eq)]
pub enum ColorSpace {
    /// BGRA8 -> NV12, BT.709 limited range (SDR)
    Rec709Sdr,
    /// R16G16B16A16Float -> P010, BT.2020 PQ (HDR10)
    Rec2020Pq,
}

pub struct VideoProcessor {
    video_device: ID3D11VideoDevice,
    video_context: ID3D11VideoContext1,
    enumerator: ID3D11VideoProcessorEnumerator,
    processor: ID3D11VideoProcessor,
    width: u32,
    height: u32,
}

// SAFETY: D3D11 video processor with multithread-protected context.
unsafe impl Send for VideoProcessor {}
unsafe impl Sync for VideoProcessor {}

impl VideoProcessor {
    pub fn new(
        device: &ID3D11Device5,
        context: &ID3D11DeviceContext4,
        width: u32,
        height: u32,
        _output_format: DXGI_FORMAT,
        cs: ColorSpace,
    ) -> WincapResult<Self> {
        let video_device: ID3D11VideoDevice = hr_call!("video_processor", device.cast());
        let video_context: ID3D11VideoContext1 = hr_call!("video_processor", context.cast());

        let desc = D3D11_VIDEO_PROCESSOR_CONTENT_DESC {
            InputFrameFormat: D3D11_VIDEO_FRAME_FORMAT_PROGRESSIVE,
            InputFrameRate: DXGI_RATIONAL {
                Numerator: 60,
                Denominator: 1,
            },
            InputWidth: width,
            InputHeight: height,
            OutputFrameRate: DXGI_RATIONAL {
                Numerator: 60,
                Denominator: 1,
            },
            OutputWidth: width,
            OutputHeight: height,
            Usage: D3D11_VIDEO_USAGE_PLAYBACK_NORMAL,
        };

        let enumerator: ID3D11VideoProcessorEnumerator =
            hr_call!("video_processor", unsafe { video_device.CreateVideoProcessorEnumerator(&desc) });
        let processor: ID3D11VideoProcessor =
            hr_call!("video_processor", unsafe { video_device.CreateVideoProcessor(&enumerator, 0) });

        if cs == ColorSpace::Rec709Sdr {
            let _in_cs = D3D11_VIDEO_PROCESSOR_COLOR_SPACE {
                _bitfield: 0, // RGB_Range=0 (full), YCbCr_Matrix=1 (BT.709)
            };
            // Set bits: Usage=0, RGB_Range=0, YCbCr_Matrix=1, YCbCr_xvYCC=0, Nominal_Range=0_255
            let mut in_cs_val = D3D11_VIDEO_PROCESSOR_COLOR_SPACE::default();
            in_cs_val._bitfield = 0b0_0_01_0_01; // YCbCr_Matrix=1, Nominal_Range=0_255
            unsafe {
                video_context.VideoProcessorSetStreamColorSpace(&processor, 0, &in_cs_val);
            }

            let mut out_cs = D3D11_VIDEO_PROCESSOR_COLOR_SPACE::default();
            out_cs._bitfield = 0b0_0_01_0_10; // YCbCr_Matrix=1, Nominal_Range=16_235
            unsafe {
                video_context.VideoProcessorSetOutputColorSpace(&processor, &out_cs);
            }
        } else {
            // HDR10: scRGB linear float input -> BT.2020 PQ studio-range YUV.
            // Use the DXGI_COLOR_SPACE path via VideoContext2 if available.
            if let Ok(ctx2) = video_context.cast::<ID3D11VideoContext2>() {
                use windows::Win32::Graphics::Dxgi::Common::*;
                unsafe {
                    ctx2.VideoProcessorSetStreamColorSpace1(
                        &processor,
                        0,
                        DXGI_COLOR_SPACE_RGB_FULL_G10_NONE_P709, // scRGB
                    );
                    ctx2.VideoProcessorSetOutputColorSpace1(
                        &processor,
                        DXGI_COLOR_SPACE_YCBCR_STUDIO_G2084_LEFT_P2020, // BT.2020 PQ
                    );
                }
            } else {
                // Fallback for older systems.
                let mut in_cs = D3D11_VIDEO_PROCESSOR_COLOR_SPACE::default();
                in_cs._bitfield = 0b0_0_00_0_01; // Full range
                unsafe {
                    video_context.VideoProcessorSetStreamColorSpace(&processor, 0, &in_cs);
                }
                let mut out_cs = D3D11_VIDEO_PROCESSOR_COLOR_SPACE::default();
                out_cs._bitfield = 0b0_0_00_0_10; // 16-235
                unsafe {
                    video_context.VideoProcessorSetOutputColorSpace(&processor, &out_cs);
                }
            }
        }

        let full = RECT {
            left: 0,
            top: 0,
            right: width as i32,
            bottom: height as i32,
        };
        unsafe {
            video_context.VideoProcessorSetStreamSourceRect(&processor, 0, true, Some(&full));
            video_context.VideoProcessorSetStreamDestRect(&processor, 0, true, Some(&full));
            video_context.VideoProcessorSetOutputTargetRect(&processor, true, Some(&full));
        }

        Ok(Self {
            video_device,
            video_context,
            enumerator,
            processor,
            width,
            height,
        })
    }

    /// Convert one frame. `dest` must have the matching output format.
    pub fn convert(&self, src: &ID3D11Texture2D, dest: &ID3D11Texture2D) -> WincapResult<()> {
        let in_desc = D3D11_VIDEO_PROCESSOR_INPUT_VIEW_DESC {
            FourCC: 0,
            ViewDimension: D3D11_VPIV_DIMENSION_TEXTURE2D,
            Anonymous: D3D11_VIDEO_PROCESSOR_INPUT_VIEW_DESC_0 {
                Texture2D: D3D11_TEX2D_VPIV { MipSlice: 0, ArraySlice: 0 },
            },
        };
        let mut in_view: Option<ID3D11VideoProcessorInputView> = None;
        hr_call!("video_processor", unsafe {
            self.video_device
                .CreateVideoProcessorInputView(src, &self.enumerator, &in_desc, Some(&mut in_view))
        });
        let in_view = in_view.ok_or_else(|| crate::error::WincapError::General {
            component: "video_processor",
            message: "CreateVideoProcessorInputView returned None".into(),
        })?;

        let out_desc = D3D11_VIDEO_PROCESSOR_OUTPUT_VIEW_DESC {
            ViewDimension: D3D11_VPOV_DIMENSION_TEXTURE2D,
            Anonymous: D3D11_VIDEO_PROCESSOR_OUTPUT_VIEW_DESC_0 {
                Texture2D: D3D11_TEX2D_VPOV { MipSlice: 0 },
            },
        };
        let mut out_view: Option<ID3D11VideoProcessorOutputView> = None;
        hr_call!("video_processor", unsafe {
            self.video_device
                .CreateVideoProcessorOutputView(dest, &self.enumerator, &out_desc, Some(&mut out_view))
        });
        let out_view = out_view.ok_or_else(|| crate::error::WincapError::General {
            component: "video_processor",
            message: "CreateVideoProcessorOutputView returned None".into(),
        })?;

        let stream = D3D11_VIDEO_PROCESSOR_STREAM {
            Enable: true.into(),
            OutputIndex: 0,
            InputFrameOrField: 0,
            PastFrames: 0,
            FutureFrames: 0,
            pInputSurface: unsafe { std::mem::transmute_copy(&in_view) },
            ..Default::default()
        };

        hr_call!("video_processor", unsafe {
            self.video_context
                .VideoProcessorBlt(&self.processor, &out_view, 0, &[stream])
        });

        Ok(())
    }

    pub fn width(&self) -> u32 {
        self.width
    }

    pub fn height(&self) -> u32 {
        self.height
    }
}
